authentication/app.js

```diff
@@ -0,0 +1,259 @@
+const FIELDAGENT_OAUTH_TOKEN_PATH = '/oauth/token'
+const FIELDAGENT_GRAPHQL_PATH = '/graphql'
+
+const INDEX_HTML = 'index.html'
+const FIELDS_HTML = 'fields.html'
+const NOT_FOUND_HTML = '404.html'
+
+const requiredEnvVars = [
+  { name: 'CLIENT_ID', value: CLIENT_ID },
+  { name: 'CLIENT_SECRET', value: CLIENT_SECRET }
+]
+requiredEnvVars.forEach(envVar => {
+  if (!envVar.value) {
+    const message = `${envVar.name} environment variable not defined. Request this value from Sentera by sending an email to api-support@sentera.com.`
+    console.log(message)
+    process.exit(1)
+  }
+})
+
+const fieldAgent = FIELDAGENT_PROTOCOL === 'https' ? https : http
+
+const requestListener = (clientRequest, clientResponse) => {
+  // Process requests made by this demo app's web UI
+  clientResponse.setHeader("Content-Type", "text/html")
+
+  const clientRequestInfo = url.parse(clientRequest.url, true)
+  const path = clientRequestInfo.pathname
+  console.log(`Process ${path} request`)
+
+  switch (path) {
+    case '/':
+      handleIndexRequest(clientResponse)
+      break
+
+    case '/authenticate':
+      handleAuthenticationRequest(clientResponse)
+      break
+
+    case '/authenticate/callback':
+      handleAuthenticationCallback(clientRequestInfo, clientResponse)
+      break
+
+    default:
+      // Requested path does not exist
+      clientResponse.writeHead(404)
+      const html = htmlFiles[NOT_FOUND_HTML].replace('%{path}', path)
+      clientResponse.end(html)
+  }
+}
+
+const server = http.createServer(requestListener)
+
+// Read the HTML files used by this app into memory,
+// and then start the server
+
+const htmlFiles = {}
+const htmlFileNames = [INDEX_HTML, FIELDS_HTML, NOT_FOUND_HTML]
+const promises = htmlFileNames.map(filename => (
+  fs.readFile(`${__dirname}/${filename}`)
+    .then(contents => {
+      htmlFiles[filename] = contents.toString()
+    })
+    .catch(error => {
+      console.error(`Could not read ${filename}: ${error}`)
+      process.exit(1)
+    })
+))
+
+Promise.all(promises).then(() => {
+  server.listen(APP_PORT, APP_HOST, () => {
+    console.log(`Server is running on ${APP_BASE_URL}`)
+  })
+})
+
+// Request Handler Functions
+
+const handleIndexRequest = (clientResponse) => {
+  clientResponse.writeHead(200)
+  const html = htmlFiles[INDEX_HTML].replace('%{error}', '')
+  clientResponse.end(html)
+}
+
+const handleAuthenticationRequest = (clientResponse) => {
+  // To retrieve fields from FieldAgent, we need to get an
+  // access token so we can call the FieldAgent GraphQL API.
+  // Initiate the OAuth authentication code grant workflow
+  // by redirecting the browser to FieldAgent for authentication.
+  const params = new URLSearchParams({
+    'response_type': 'code',
+    'client_id': CLIENT_ID,
+    'scope': CLIENT_SCOPE,
+    'state': CLIENT_STATE,
+    'redirect_uri': APP_CALLBACK_URI
+  }).toString();
+  clientResponse.writeHead(302, {
+    'Location': FIELDAGENT_AUTH_URL + '?' + params
```

Comment:
```suggestion
    response_type: 'code',
    client_id: CLIENT_ID,
    scope: CLIENT_SCOPE,
    state: CLIENT_STATE,
    redirect_uri: APP_CALLBACK_URI
  }).toString();
  clientResponse.writeHead(302, {
    Location: FIELDAGENT_AUTH_URL + '?' + params
```

Reply:
Might be worth running this through `standard --fix` real quick to tidy things up, even if this repo doesn't include any packages or enforce any linting rules. Looks like standard would clean up some inconsistencies regarding quotes, semicolons, and `const`/`var` declarations.

Reply:
Good point. I will do that.

Reply:
Done! https://github.com/SenteraLLC/code-examples/pull/1/commits/baedba4609101d4ed052b69db06524d2dae81de7

Reply:
Very standard!